model/protecode.py

```diff
@@ -68,3 +89,45 @@ def _required_attributes(self):
     def validate(self):
         # call credentials method to validate credentials
         self.credentials()
+
+
+def find_config(
+    group_id: int,
+    base_url: str,
+    cfg_factory=None,
+) -> ProtecodeConfig | None:
+    if not cfg_factory:
+        cfg_factory = ci.util.ctx().cfg_factory()
+
+    matching_cfgs = []
+    hostname = ci.util.urlparse(base_url).hostname
+
+    for cfg in cfg_factory._cfg_elements(cfg_type_name='protecode'):
+        cfg: ProtecodeConfig
+
+        if cfg.hostname() and not cfg.matches_hostname(hostname):
+            continue
+
+        if cfg.group_ids() and not cfg.matches_group_id(group_id):
+            continue
+
+        matching_cfgs.append(cfg)
+
+    if not matching_cfgs:
+        return None
+
+    if [
+        cfg.hostname()
+        for cfg in matching_cfgs
+    ]:
+        # eliminate less specific cfgs
+        matching_cfgs = [
+            cfg
+            for cfg in matching_cfgs
+            if cfg.hostname()
+        ]
+
+    matching_cfgs = sorted(matching_cfgs, key=lambda cfg: len(cfg.group_ids()))
```

Comment:
why would you sort by amount of configured group-IDs? if at all, less group-ids would be more specific (however, it does not make a lot of sense to sort by smaller number of group-ids as well).

a metric I would consider for measuring how (relatively) specific a given bdba-cfg is would be:
- does it, or does not not define a base-url (if not, consider it less specific than a bdba-cfg that does)
- does it, or does it not define group-ids (if not, consider it less specific than a bdba-cfg that does)

a lookup-semantics working on this could look like this:
- discard cfg-elements that do have an explicit cfg, but have cfg-values that do not match the desired values (examples: 1. the base-url does not match the passed value for base-url, 2. the group-ids do not match the passed-in value(s) for group-id)
- out of the remaining cfg-elements (i.e. those that did not formally disqualify), order cfg-elements by how specific they are
  - since each cfg will have a base-url, this attribute might actually be ignored (if base-url would not match, the cfg-element would have been discarded at this point; if the cfg-element does not have a base-url, it is invalid (note: please referain from actually adding such a check here)
  - which leaves the group-id. check whether there is a cfg-element that does declare group-ids. take e.g. the first one that has. fallback: take a bdba-cfg that does not declare group-ids